eth/watchers/roundswatcher.go

```diff
@@ -0,0 +1,160 @@
+package watchers
+
+import (
+	"fmt"
+	"math/big"
+	"sync"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/livepeer/go-livepeer/eth/contracts"
+
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/eth"
+	"github.com/livepeer/go-livepeer/eth/blockwatch"
+)
+
+// RoundsWatcher is a type alias for a thread safe in-memory cache that watches on-chain events for new initialized rounds
+type RoundsWatcher struct {
+	mu                   sync.RWMutex
+	lastInitializedRound *big.Int
+	blockHashForRound    [32]byte
+
+	quit chan struct{}
+
+	watcher Watcher
+	lpEth   eth.LivepeerEthClient
+	dec     *EventDecoder
+}
+
+type Watcher interface {
+	Subscribe(sink chan<- []*blockwatch.Event) event.Subscription
+}
+
+// NewRoundsWatcher creates a new instance of RoundsWatcher and sets the initial cache through an RPC call to an ethereum node
+func NewRoundsWatcher(watcher Watcher, lpEth eth.LivepeerEthClient, addr ethcommon.Address) (*RoundsWatcher, error) {
+	dec, err := NewEventDecoder(addr, contracts.RoundsManagerABI)
+	if err != nil {
+		return nil, fmt.Errorf("error creating decoder: %v", err)
+	}
+
+	return &RoundsWatcher{
+		quit:    make(chan struct{}),
+		watcher: watcher,
+		lpEth:   lpEth,
+		dec:     dec,
+	}, nil
+}
+
+// LastInitializedRound gets the last initialized from cache
+func (rw *RoundsWatcher) LastInitializedRound() (*big.Int, error) {
+	rw.mu.RLock()
+	defer rw.mu.RUnlock()
+	return rw.lastInitializedRound, nil
+}
+
+// BlockHashForRound gets the blockhash for the last initialized round from cache
+func (rw *RoundsWatcher) BlockHashForRound(round *big.Int) ([32]byte, error) {
+	if round == nil {
+		return [32]byte{}, fmt.Errorf("round is nil")
+	}
+	lastRound, _ := rw.LastInitializedRound()
+	if lastRound.Cmp(round) != 0 {
+		return rw.lpEth.BlockHashForRound(round)
+	}
+	rw.mu.RLock()
+	defer rw.mu.RUnlock()
+	return rw.blockHashForRound, nil
+}
+
+// setLastInitializedRound sets the last initialized round in cache
+func (rw *RoundsWatcher) setLastInitializedRound(round *big.Int) {
+	rw.mu.RLock()
+	defer rw.mu.RUnlock()
+	rw.lastInitializedRound = round
+}
+
+//setBlochashForRound sets the blockhash for the last initialized round in cache
+func (rw *RoundsWatcher) setBlockHashForRound(hash [32]byte) {
+	rw.mu.RLock()
+	defer rw.mu.RUnlock()
+	rw.blockHashForRound = hash
+}
+
+// Watch the blockwatch subscription for NewRound events
+func (rw *RoundsWatcher) Watch() {
+	lr, err := rw.lpEth.LastInitializedRound()
+	if err != nil {
+		glog.Errorf("error fetching initial lastInitializedRound value: %v", err)
+	}
+	rw.setLastInitializedRound(lr)
+
+	bh, err := rw.lpEth.BlockHashForRound(lr)
+	if err != nil {
+		glog.Errorf("error fetching initial blockHashForRound value: %v", err)
+	}
+	rw.setBlockHashForRound(bh)
+
+	events := make(chan []*blockwatch.Event)
```

Comment:
We should make this a buffered channel so that additional block events can be queued up even if the rounds watcher is still processing existing block events. The UnbondingWatcher currently uses a [buffer size of 10](https://github.com/livepeer/go-livepeer/pull/1048/files#diff-fd62975d7c6fadd8b06b097faddf0074R50).

Reply:
fixed in 80d3e25